visuals/two_nights.py

```diff
@@ -0,0 +1,342 @@
+        )
+
+        df["Day"] = df["Time (UTC)"].dt.day
+        df["Month"] = df["Time (UTC)"].dt.month
+        df["Year"] = df["Time (UTC)"].dt.year
+        df["Date"] = df["Time (UTC)"].apply(lambda x: x.date())
+
+    def _construct_data(self):
+        """
+        This method constructs all the information needed for the
+        class to render the visualization.
+        """
+        pass
+
+    def _get_moon_info(self, df: pd.DataFrame):
+        def get_moon_illum(t: datetime):
+            return float(str(self.sensor_location.moon_illumination(str(t))))
+
+        df["moon_illum"] = df["Time (CST)"].apply(lambda t: get_moon_illum(t))
+
+    def _call_apis(self):
+        return super()._call_apis()
+
+    def _call_weather_api(self, t_start: datetime, t_end: datetime):
+        """
+        This method pulls the relevant weather for the night
+        from external APIs.
+
+        Returns: dictionary of hours, cloud cover percentages
+        """
+        api_url = (
+            f"https://archive-api.open-meteo.com/v1/archive"
+            f"?latitude={self.sensor_latitude}"
+            f"&longitude={self.sensor_longitude}"
+            f"&start_date={t_start.date()}&end_date={t_end.date()}"
+            f"&hourly=cloudcover&timezone=America%2FChicago"
+        )
+        response = requests.get(api_url)
+        data = response.json()
+
+        if not data:
+            raise Exception("Cloud Data Couldn't be Found")
+
+        times = [
+            datetime.strptime(time, "%Y-%m-%dT%H:%M%f")
+            for time in data["hourly"]["time"]
+        ]
+        cloudcover = data["hourly"]["cloudcover"]
+
+        cloud_dict = dict(zip(times, cloudcover))
+        return {x: cloud_dict[x] for x in cloud_dict if t_start <= x <= t_end}
+
+    def _get_sunset_sunrise(self, t_start: datetime, t_end: datetime):
+        sunset_mjd_time = (
+            float(
+                str(
+                    self.sensor_location.sun_set_time(
+                        Time(str(t_start)), which="nearest"
+                    )
+                )
+            )
+            - 2400000.5
+            - (6 * (0.5 / 12))
+        )
+        sunset_standard_time = julian.from_jd(sunset_mjd_time, fmt="mjd")
+        sunset_24time = sunset_standard_time.strftime(("%H:%M"))
+
+        sunrise_mjd_time = (
+            float(
+                str(
+                    self.sensor_location.sun_rise_time(
+                        Time(str(t_end)), which="nearest"
+                    )
+                )
+            )
+            - 2400000.5
+            - (6 * (0.5 / 12))
+        )
+        sunrise_standard_time = julian.from_jd(sunrise_mjd_time, fmt="mjd")
+        sunrise_24time = sunrise_standard_time.strftime(("%H:%M"))
+
+        return sunset_24time, sunrise_24time
+
+    def create_visual(
+        self,
+        t_start1: datetime,
+        t_end1: datetime,
+        t_start2: datetime,
+        t_end2: datetime,
+    ):
```

Comment:
This does not match the `create_visual` function from the abstract class. Modify one of them so that it is correct.